---
title: "check overlapping variants"
author: "Daan Hazelaar"
date: "`r Sys.Date()`"
output: pdf_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(VariantAnnotation)
library(SeqKat)
source(file = "../R/function_package_workflows.R")
source(file = "../R/runSigProfiler.R")
```

Manual generation of a VRanges object that contains 10 overlapping variants
```{r}
startLocationsRandomSNVs <- c(1003484, 1007628, 1044954, 1052411, 1062310, 1064989, 1073941, 1082478, 1084582, 1094532)
endLocationsRandomSNVs <- startLocationsRandomSNVs

startLocationKataegisSNVs <- c(1050178, 1050184, 1050187, 1050421, 1050430, 1050441, 1050510, 1050632, 1050877, 1050981)
endLocationKataegisSNVs <- startLocationKataegisSNVs

startLocationINDEL <- 1050800
endLocationINDEL <- 1052000

startlocationOverlappingSNVs <- c(1003484, 1007628, 1050178, 1050184, 1020000, 1020000, 1020000, 1020005, 1020005)
endlocationOverlappingSNVs <- startlocationOverlappingSNVs

testTib <- tibble(
    sampleNames = "testSyntheticOverlapSample",
    seqnames = "chr1",
    start = c(startLocationsRandomSNVs, startLocationKataegisSNVs, startLocationINDEL, startlocationOverlappingSNVs),
    end = c(endLocationsRandomSNVs, endLocationKataegisSNVs, endLocationINDEL, endlocationOverlappingSNVs)
) |> 
    mutate(
        ref =  BSgenome::getSeq(
            BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19,
            names = "chr1",
            start = start,
            end = end,
            as.character = TRUE),
        alt = base::vapply(
            ref,
            FUN = function(x){
                nucleotides <- c("A", "T", "C", "G")
                base::sample(nucleotides[nucleotides != x], size = 1)
            }, FUN.VALUE = 's')
    ) 

testVR <- testTib |> 
    GenomicRanges::makeGRangesFromDataFrame(keep.extra.columns = TRUE) |> 
    VariantAnnotation::makeVRangesFromGRanges() |> 
    GenomicRanges::sort()

testMAF <- testVR |> 
    tibble::as_tibble() |>
    dplyr::mutate(
        Hugo_Symbol = 'Unknown',
        Variant_Classification = 'Missense_Mutation',
        Variant_Type = 'SNP',
    ) |>
    dplyr::select(
        Hugo_Symbol,
        Chromosome = .data$seqnames,
        Start_Position = .data$start,
        End_Position = .data$end,
        Reference_Allele = .data$ref,
        Tumor_Seq_Allele2 = .data$alt,
        Variant_Classification,
        Variant_Type,
        Tumor_Sample_Barcode = .data$sampleNames
    )
```

Run packages
```{r}
kd <- katdetectr::detectKataegis(genomicVariants = testVR)
mf <- runMaftools(genomicVariants = testMAF)
sq <- runSeqkat(genomicVariants = testVR)
ka <- runKataegis(genomicVariants = testVR)
cm <- runClusteredMutations(genomicVariants = testVR)
sig <- runSigProfilerClustered(data = testVR)
```

```{r}
data <- testVR

sampleName <- unique(data@sampleNames)

# create new dir
dir.create(path = paste0("", sampleName), showWarnings = FALSE)
setwd(dir = paste0("", sampleName))

# convert data to a sigprofiler friendly format
GenomeInfoDb::seqlevelsStyle(data) <- 'NCBI'
vrFormatted <- data |> tibble::as_tibble() |>
    mutate(
        CancerType = ".",
        referenceGenome = "GRCh37",
        NGS_method = ".",
        mutss = "SOMATIC",
        mutType = "SNV"
    ) |>
    dplyr::select(CancerType, sampleNames, NGS_method, referenceGenome, mutType, seqnames, start, end, ref, alt, mutss)

# write data to disk as a .txt file
write.table(vrFormatted, file = "data_SigProf.txt", row.names = FALSE, col.names = FALSE, quote = FALSE, sep = "\t")
# add the header to the data.txt
file.copy("data/header.txt", "data_SigProf_header.txt", overwrite = TRUE)
file.append("data_SigProf_header.txt", "data_SigProf.txt")
unlink("data_SigProf.txt")


startTime <- base::proc.time()
#system("python ../../prunSigProfiler.py")
system("python3 ../python/prunSigProfiler.py")
```

